File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/profile/BinaryOperation.java
package gov.nasa.jpl.aerie.constraints.profile;

import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * A base-preserving map out of the product of two pointed sets.
 *
 * The basepoint of a product is the pair of their basepoints. Only this point needs to be sent
 * to the target set's basepoint. If one or both elements of the tuple are defined, the transformation
 * can do whatever it likes.
 */
public interface BinaryOperation<Left, Right, Out> {
  // When only one value is present
  Optional<Out> left(Left left);
  Optional<Out> right(Right right);
  // When both values are present
  Optional<Out> combine(Left left, Right right);
  // When no values are present, the transformation always produces `empty`.

  static <Left, Right, Out> BinaryOperation<Left, Right, Out> from(final BiFunction<Optional<Left>, Optional<Right>, Optional<Out>> f) {
    return new BinaryOperation<>() {
      @Override
      public Optional<Out> combine(final Left left, final Right right) {
        return f.apply(Optional.of(left), Optional.of(right));
      }

      @Override
      public Optional<Out> left(final Left left) {
        return f.apply(Optional.of(left), Optional.empty());
      }

      @Override
      public Optional<Out> right(final Right right) {
        return f.apply(Optional.empty(), Optional.of(right));
      }
    };
  }

  static <Left, Right, Out> BinaryOperation<Left, Right, Out> fromCases(
      final Function<Left, Optional<Out>> left,
      final Function<Right, Optional<Out>> right,
      final BiFunction<Left, Right, Optional<Out>> combine
  ) {
    return new BinaryOperation<>() {
      @Override
      public Optional<Out> combine(final Left a, final Right b) {
        return combine.apply(a, b);
      }

      @Override
      public Optional<Out> left(final Left a) {
        return left.apply(a);
      }

      @Override
      public Optional<Out> right(final Right b) {
        return right.apply(b);
      }
    };
  }

  static <Left, Right, Out> BinaryOperation<Left, Right, Out> combineDefaultEmpty(final BiFunction<Left, Right, Out> f) {
    return new BinaryOperation<>() {
      @Override
      public Optional<Out> left(final Left left) {
        return Optional.empty();
      }

      @Override
      public Optional<Out> right(final Right right) {
        return Optional.empty();
      }

      @Override
      public Optional<Out> combine(final Left left, final Right right) {
        return Optional.of(f.apply(left, right));
      }
    };
  }

  static <V> BinaryOperation<V, V, V> combineDefaultIdentity(final BiFunction<V, V, V> f) {
    return new BinaryOperation<>() {
      @Override
      public Optional<V> left(final V left) {
        return Optional.of(left);
      }

      @Override
      public Optional<V> right(final V right) {
        return Optional.of(right);
      }

      @Override
      public Optional<V> combine(final V left, final V right) {
        return Optional.of(f.apply(left, right));
      }
    };
  }

  enum OpMode {
    Left, Right, Combine
  }
}

File: merlin-server/constraints-dsl-compiler/src/libs/binary-operation.ts
export class BinaryOperation<Left, Right, Out> {
  public readonly left: (l: Left) => Out | undefined;
  public readonly right: (r: Right) => Out | undefined;
  public readonly combine: (l: Left, r: Right) => Out | undefined;

  private constructor(
      left: (l: Left) => Out | undefined,
      right: (r: Right) => Out | undefined,
      combine: (l: Left, r: Right) => Out | undefined
  ) {
    this.left = left;
    this.right = right;
    this.combine = combine;
  }

  public static cases<Left, Right, Out>(
      left: (l: Left) => Out | undefined,
      right: (r: Right) => Out | undefined,
      combine: (l: Left, r: Right) => Out | undefined
  ): BinaryOperation<Left, Right, Out> {
    return new BinaryOperation<Left, Right, Out>(left, right, combine);
  }

  public static singleFunction<Left, Right, Out>(
      func: (l: Left | undefined, r: Right | undefined) => Out | undefined
  ): BinaryOperation<Left, Right, Out> {
    return new BinaryOperation(
        l => func(l, undefined),
        r => func(undefined, r),
        (l, r) => func(l, r)
    );
  }

  public static combineOrUndefined<Left, Right, Out>(
      func: (l: Left, r: Right) => Out | undefined
  ): BinaryOperation<Left, Right, Out> {
    return BinaryOperation.cases(
        _ => undefined,
        _ => undefined,
        (l, r) => func(l, r)
    );
  }

  public static combineOrIdentity<Left, Right, Out>(
      func: (l: Left, r: Right) => Out | undefined
  ): BinaryOperation<Left, Right, Out> {
    return BinaryOperation.cases(
        l => l,
        r => r,
        (l, r) => func(l, r)
    );
  }
}

export enum OpMode {
  Left,
  Right,
  Combine
}
